### PR TITLE
Customizable channel ConcatAll

### PR DESCRIPTION
**Pull request by vigoo**
The motivation of this is to provide low level tools to implement zio-test's `flatMapStream` more efficiently.

Benchmark results:

```
Gen.listOfN(size)(Gen.byte).sample.forever.collectSome.take(count).runDrain

old flatMapStream:
[info] Benchmark             (count)  (size)   Mode  Cnt  Score   Error  Units
[info] GenBenchmark.listOfN       10      10  thrpt   10  326,355 ± 11,690  ops/s
[info] GenBenchmark.listOfN       10     100  thrpt   10  19,433  ± 0,301  ops/s
[info] GenBenchmark.listOfN      100    1000  thrpt   10  0,030   ± 0,007  ops/s

[info] GenBenchmark.causes        10      10  thrpt   10  258,529 ±  9,882  ops/s
[info] GenBenchmark.causes        10     100  thrpt   10  20,666  ± 1,162  ops/s

new flatMapStream:
[info] Benchmark             (count)  (size)   Mode  Cnt  Score   Error  Units
[info] GenBenchmark.listOfN       10      10  thrpt   10  624,059 ± 40,024  ops/s
[info] GenBenchmark.listOfN       10     100  thrpt   10  52,105  ± 1,349  ops/s
[info] GenBenchmark.listOfN      100    1000  thrpt   10  0,466   ± 0,003  ops/s

[info] GenBenchmark.causes        10      10  thrpt   10  420,012 ± 16,113  ops/s
[info] GenBenchmark.causes        10     100  thrpt   10  42,814  ± 1,917  ops/s

after optimizations merged on 22/12/2021:
[info] GenBenchmark.listOfN       10      10  thrpt   10  751,595 ± 19,347  ops/s
[info] GenBenchmark.listOfN       10     100  thrpt   10  59,483  ± 1,029  ops/s
[info] GenBenchmark.listOfN      100    1000  thrpt   10  0,525   ± 0,013  ops/s

[info] GenBenchmark.causes        10      10  thrpt   10  498,143 ±  6,898  ops/s
[info] GenBenchmark.causes        10     100  thrpt   10  50,752  ± 1,122  ops/s
```

It would be possible to add more customization possibilities by extending `ChildExecutorDecision`, I only did the minimum necessary.

### CODE REVIEW

**Comment by vigoo** [comment]
Needs https://github.com/zio/zio/pull/6127

File: test/shared/src/main/scala/zio/test/package.scala [hunk 1]
```diff
@@ -1013,174 +1015,56 @@ package object test extends CompileVariants {
...
-          (Some(builder.result()), innerChunk, i)
-        }
-
-      currentInnerStream.get.flatMap {
-        case None =>
-          currentStreams.get.map(_.headOption).flatMap {
+  )(f: A => ZStream[R1, Nothing, Option[B]])(implicit trace: ZTraceElement): ZStream[R1, Nothing, Option[B]] =
+    new ZStream(
+      stream
+        .rechunk(1)
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/package.scala:1023]: Is this necessary?

Comment (vigoo) [reply test/shared/src/main/scala/zio/test/package.scala:1023]: Currently the "customization" only allows to switch streams after each emit, so if there would be a `None` in the middle of the chunk we could not split it and only emit the first half and store the rest for a next run.

For that we would have to add the ability to also switch where we are reading from, to be able to continue from a buffered remainder of the chunk, and also probably to make the `onEmit` an effect.

Comment (vigoo) [reply test/shared/src/main/scala/zio/test/package.scala:1023]: Maybe we could also "dynamically" chunk it at each `None`?

Comment (adamgfraser) [reply test/shared/src/main/scala/zio/test/package.scala:1023]: Sure.

**Review by adamgfraser — APPROVED** [review]

**Pull request merged**